akka-actor/src/main/scala/akka/pattern/PromiseRef.scala

```diff
@@ -0,0 +1,195 @@
+
+    /**
+     * Provides a conversion to transparently use PromiseRef as ActorRef.
+     */
+
+    implicit def toActorRef(promiseRef: PromiseRef[_]): ActorRef = promiseRef.ref
+
+    /**
+     * Provides a conversion to transparently use PromiseRef to dispatch messages via Scala DSL.
+     */
+    implicit def toScalaActorRef(promiseRef: PromiseRef[_]): ScalaActorRef = promiseRef.ref
+  }
+}
+
+object FutureRef {
+  /**
+   * Wraps an ActorRef and a Future into a FutureRef.
+   */
+  private[akka] def wrap[T](actorRef: ActorRef, future: Future[T]): FutureRef[T] = {
+    new FutureRefImpl(actorRef, future)
+  }
+
+  /**
+   * Constructs a new FutureRef which will be completed with the first message sent to it.
+   *
+   * {{{
+   * // enables transparent use of FutureRef as ActorRef and Future
+   * import FutureRef.Implicits._
+   *
+   * val futureRef = FutureRef(system, 5.seconds)
+   * futureRef ! "message"
+   * futureRef.onComplete(println)  // prints "message"
+   * }}}
+   */
+  def apply(system: ActorSystem, timeout: Timeout): FutureRef[Any] = {
+    PromiseRef(system, timeout).toFutureRef
+  }
+
+  /**
+   * Constructs a new PromiseRef which will be completed with the first message sent to it.
+   *
+   * {{{
+   * // enables transparent use of FutureRef as ActorRef and Promise
+   * import FutureRef.Implicits._
+   *
+   * // requires an implicit ActorSystem in scope
+   * val futureRef = FutureRef(5.seconds)
+   * futureRef ! "message"
+   * futureRef.onComplete(println)  // prints "message"
+   * }}}
+   */
+  def apply(timeout: Timeout)(implicit system: ActorSystem): FutureRef[Any] = {
+    FutureRef(system, timeout)
+  }
+
+  object Implicits {
+    /**
+     * Provides a conversion to transparently use FutureRef as Future.
+     */
+    implicit def toFuture[T](futureRef: FutureRef[T]): Future[T] = futureRef.future
+
+    /**
+     * Provides a conversion to transparently use FutureRef as ActorRef.
+     */
+    implicit def toScalaActorRef(futureRef: FutureRef[_]): ScalaActorRef = futureRef.ref
+
+    /**
+     * Provides a conversion to transparently use FutureRef to dispatch messages via Scala DSL.
+     */
+    implicit def toActorRef(futureRef: FutureRef[_]): ActorRef = futureRef.ref
+  }
+}
+
+private[akka] class PromiseRefImpl[T](val ref: ActorRef, val promise: Promise[T])
+  extends PromiseRef[T] with FutureRef[T] {
+  def toFutureRef: FutureRef[T] = this
+}
+
+private[akka] final class FutureRefImpl[T](val ref: ActorRef, val future: Future[T])
+  extends FutureRef[T]
+
+private[akka] final class AskPromiseRef private (promiseActorRef: PromiseActorRef)
+  extends PromiseRefImpl[Any](promiseActorRef, promiseActorRef.result)
+
+private[akka] object AskPromiseRef {
+  def apply(provider: ActorRefProvider, timeout: Timeout): AskPromiseRef = {
+    if (timeout.duration.length > 0) {
+      val promiseActorRef = PromiseActorRef(provider, timeout, "unknown", "unknown", ActorRef.noSender)
```

Comment:
please use `provider.deadLetters` instead of `null` (which is what `noSender` means)